File: Algorithms/src/2996. Smallest Missing Integer Greater Than Sequential Prefix Sum/Code/solution.go
func missingInteger(nums []int) int {
    // I start the sum with the first element because
    // nums[0] is always part of the sequential prefix.
    sum := nums[0]

    // I scan from the second element to find the end of the sequential prefix.
    for i := 1; i < len(nums); i++ {
        // The sequence continues only when the current value
        // is exactly one greater than the previous value.
        if nums[i] == nums[i-1]+1 {
            sum += nums[i]
        } else {
            // The sequence breaks here, so I stop the prefix scan.
            break
        }
    }

    // I use a map as a set to store every value from the array.
    seen := make(map[int]bool)

    // I mark every array value as present.
    for _, num := range nums {
        seen[num] = true
    }

    // I start checking candidates from the sequential prefix sum.
    answer := sum

    // If the candidate exists, I keep increasing it until
    // I find a number that is missing from the array.
    for seen[answer] {
        answer++
    }

    // The first missing candidate is the required answer.
    return answer
}

File: Algorithms/src/2996. Smallest Missing Integer Greater Than Sequential Prefix Sum/Code/solution.py
from typing import List

class Solution:
    def missingInteger(self, nums: List[int]) -> int:
        # I start with the first value because nums[0]
        # is always part of the sequential prefix.
        total = nums[0]

        # I scan the remaining values to find the longest sequential prefix.
        for i in range(1, len(nums)):
            # The sequence continues only if the current value
            # is exactly one greater than the previous value.
            if nums[i] == nums[i - 1] + 1:
                total += nums[i]
            else:
                # The sequence breaks here, so I stop the prefix scan.
                break

        # I create a set containing every array value for fast lookup.
        seen = set(nums)

        # I start searching from the sequential prefix sum.
        answer = total

        # If the current number exists in the array, it is not missing,
        # so I try the next number.
        while answer in seen:
            answer += 1

        # The first missing value is the required answer.
        return answer